### PR TITLE
Add deprecated directive to arguments and input values

### PR DESCRIPTION
**Pull request by smitt04**
This is a WIP and is the start for supporting `deprecated` directive in arguments and input values.

This is in relation to [RFC Proposal](https://github.com/facebook/graphql/pull/525)

### CODE REVIEW

**Comment by IvanGoncharov** [comment]
> This is a WIP and is the start for supporting deprecated directive in arguments and input values.

@smitt04 Great 🎉 Please, ping me when you will be ready for review.

**Comment by smitt04** [comment]
@IvanGoncharov I think I am ready to be reviewed.  Let me know if there are more tests that need to be implemented.

**Comment by charliecorrigan** [comment]
I'm not familiar with the graphql repo library architecture. Once this feature is added into graphql-js, will it be available in express-graphql? Or will that require a new PR to duplicate the feature?

**Comment by modosc** [comment]
i  attempted the rebase/merge [here](https://github.com/modosc/graphql-js/tree/feature/deprecation) - one spec is still failing though.

**Comment by smitt04** [comment]
I tried updating the pr

**Comment by modosc** [comment]
@smitt04  i updated my branch, all specs pass. i don't think you can merge it in because i had already rebased off of `master` - should i open up a separate pr?

File: src/type/definition.js [hunk 1]
```diff
@@ -873,6 +875,8 @@ export function argsToArgsConfig(
       description: arg.description,
       type: arg.type,
       defaultValue: arg.defaultValue,
+      deprecationReason: arg.deprecationReason,
+      isDeprecated: arg.deprecationReason != null,
       extensions: arg.extensions,
       astNode: arg.astNode,
```
Comment (leebyron) [inline src/type/definition.js]: nit: can we make sure these arguments are always in the same order?

File: src/type/definition.js [hunk 2]
```diff
@@ -1547,6 +1563,7 @@ export type GraphQLInputFieldConfig = {|
   description?: ?string,
   type: GraphQLInputType,
   defaultValue?: mixed,
+  deprecationReason?: ?string,
```
Comment (leebyron) [inline src/type/definition.js]: isDeprecated missing here?

**Comment by IvanGoncharov** [comment]
> Should getIntrospectionQuery be updated to include deprecated input fields and arguments to be consistent with the current behavior of including deprecated fields and enum values?

@jturkel Added in #2855 Thanks for pointing it here.

**Pull request closed without merging**